File: app/voice/preview.ts
// How a timbre preview is minted (ADR 0031).
//
// Gemini Live has no sample endpoint. The same thirty `voiceName`s are
// accepted by Gemini TTS over ordinary `generateContent`, which recites a
// line in that mouth and answers with PCM. The native picker plays the
// resulting clip; a call still uses Live. Style in the line is kept out of
// it so the thirty files are comparable.

import { decodeGeminiBase64V1 } from "./gemini-live.js";

/** TTS model that shares Live's thirty prebuilt voices. */
export const GEMINI_TTS_MODEL_V1 = "gemini-2.5-flash-preview-tts";

export const GEMINI_TTS_ENDPOINT_V1 =
  "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-preview-tts:generateContent";

/** The line every clip recites, so two mouths are compared on the same words. */
export const VOICE_PREVIEW_LINE_V1 = "Say: Hi. This is how I sound.";

/** TTS output, and Live's downstream rate, so the native speaker needs no resample. */
export const VOICE_PREVIEW_SAMPLE_RATE_V1 = 24_000;

/** Where the native app loads a minted clip from. */
export function voicePreviewAssetV1(voiceName: string): string {
  return `assets/voices/${voiceName}.wav`;
}

/**
 * The one `generateContent` body that recites [VOICE_PREVIEW_LINE_V1] in
 * `voiceName`. Same `speechConfig` shape Live uses for timbre.
 */
export function buildGeminiTtsPreviewRequestV1(
  voiceName: string,
): Record<string, unknown> {
  return {
    contents: [{ parts: [{ text: VOICE_PREVIEW_LINE_V1 }] }],
    generationConfig: {
      responseModalities: ["AUDIO"],
      speechConfig: {
        voiceConfig: {
          prebuiltVoiceConfig: { voiceName },
        },
      },
    },
  };
}

export class GeminiTtsPreviewError extends Error {
  override readonly name = "GeminiTtsPreviewError";
}

/**
 * How long to wait after a 429. Prefers `Retry-After`, then the model's
 * "retry in Ns" sentence, then a minute — the free-tier TTS window.
 */
export function retryDelayFromTtsErrorV1(input: {
  status: number;
  message: string;
  retryAfter?: string | null;
}): number | undefined {
  if (input.status !== 429) return undefined;
  const header = input.retryAfter?.trim();
  if (header) {
    const seconds = Number(header);
    if (Number.isFinite(seconds) && seconds >= 0) {
      return Math.ceil(seconds * 1000);
    }
  }
  const match = /retry in ([0-9.]+)\s*s/i.exec(input.message);
  if (match) {
    const seconds = Number(match[1]);
    if (Number.isFinite(seconds) && seconds >= 0) {
      return Math.ceil(seconds * 1000);
    }
  }
  return 60_000;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function sampleRateFromMimeV1(mimeType: string): number {
  const match = /rate=(\d+)/i.exec(mimeType);
  if (!match) return VOICE_PREVIEW_SAMPLE_RATE_V1;
  const rate = Number(match[1]);
  return Number.isFinite(rate) && rate > 0
    ? rate
    : VOICE_PREVIEW_SAMPLE_RATE_V1;
}

function readString(bytes: Uint8Array, offset: number, length: number): string {
  return String.fromCharCode(...bytes.subarray(offset, offset + length));
}

/**
 * PCM16 little-endian mono from a canonical WAV, or empty when the bytes
 * are not one. Used when the model answers with a container instead of raw
 * L16.
 */
export function wavToPcmV1(wav: Uint8Array): {
  pcm: Uint8Array;
  sampleRate: number;
} | null {
  if (wav.length < 44) return null;
  if (readString(wav, 0, 4) !== "RIFF" || readString(wav, 8, 4) !== "WAVE") {
    return null;
  }
  let offset = 12;
  let sampleRate = VOICE_PREVIEW_SAMPLE_RATE_V1;
  let pcm: Uint8Array | undefined;
  while (offset + 8 <= wav.length) {
    const id = readString(wav, offset, 4);
    const size = new DataView(
      wav.buffer,
      wav.byteOffset + offset + 4,
      4,
    ).getUint32(0, true);
    const start = offset + 8;
    const end = Math.min(start + size, wav.length);
    if (id === "fmt " && end - start >= 16) {
      const view = new DataView(
        wav.buffer,
        wav.byteOffset + start,
        end - start,
      );
      sampleRate = view.getUint32(4, true) || sampleRate;
    } else if (id === "data") {
      pcm = wav.subarray(start, end);
    }
    offset = start + size + (size % 2);
  }
  if (!pcm) return null;
  return { pcm, sampleRate };
}

/** A canonical PCM16 mono WAV, the shape the native decoder expects. */
export function pcmToWavV1(
  pcm: Uint8Array,
  sampleRate: number = VOICE_PREVIEW_SAMPLE_RATE_V1,
): Uint8Array {
  const header = new ArrayBuffer(44);
  const view = new DataView(header);
  const write = (offset: number, text: string) => {
    for (let index = 0; index < text.length; index += 1) {
      view.setUint8(offset + index, text.charCodeAt(index));
    }
  };
  write(0, "RIFF");
  view.setUint32(4, 36 + pcm.byteLength, true);
  write(8, "WAVE");
  write(12, "fmt ");
  view.setUint32(16, 16, true);
  view.setUint16(20, 1, true);
  view.setUint16(22, 1, true);
  view.setUint32(24, sampleRate, true);
  view.setUint32(28, sampleRate * 2, true);
  view.setUint16(32, 2, true);
  view.setUint16(34, 16, true);
  write(36, "data");
  view.setUint32(40, pcm.byteLength, true);
  const out = new Uint8Array(44 + pcm.byteLength);
  out.set(new Uint8Array(header), 0);
  out.set(pcm, 44);
  return out;
}

/**
 * The audio part of a TTS `generateContent` response, rewritten as a
 * canonical WAV. An error payload or a missing part throws.
 */
export function wavFromGeminiTtsResponseV1(payload: unknown): Uint8Array {
  if (!isRecord(payload)) {
    throw new GeminiTtsPreviewError("TTS response must be an object");
  }
  if (isRecord(payload.error)) {
    const message =
      typeof payload.error.message === "string" && payload.error.message
        ? payload.error.message
        : "TTS request failed";
    throw new GeminiTtsPreviewError(message);
  }
  const candidates = payload.candidates;
  if (!Array.isArray(candidates) || candidates.length === 0) {
    throw new GeminiTtsPreviewError("TTS response has no candidates");
  }
  const content = isRecord(candidates[0]) ? candidates[0].content : undefined;
  const parts = isRecord(content) ? content.parts : undefined;
  if (!Array.isArray(parts)) {
    throw new GeminiTtsPreviewError("TTS response has no audio part");
  }
  for (const part of parts) {
    if (!isRecord(part) || !isRecord(part.inlineData)) continue;
    const data = part.inlineData.data;
    if (typeof data !== "string" || !data) continue;
    const mime =
      typeof part.inlineData.mimeType === "string"
        ? part.inlineData.mimeType
        : "";
    const bytes = decodeGeminiBase64V1(data);
    if (bytes.length === 0) {
      throw new GeminiTtsPreviewError("TTS audio was empty");
    }
    const wav = wavToPcmV1(bytes);
    if (wav) return pcmToWavV1(wav.pcm, wav.sampleRate);
    return pcmToWavV1(bytes, sampleRateFromMimeV1(mime));
  }
  throw new GeminiTtsPreviewError("TTS response has no audio part");
}

File: apps/native/lib/voice/preview.dart
/// Timbre preview clips minted from Gemini TTS, played through the same
/// speaker a call uses.
///
/// Live has no sample endpoint. `scripts/mint-voice-previews.ts` recites one
/// line in each of the thirty `voiceName`s over HTTP and writes a WAV under
/// `assets/voices/`. This module loads that file, strips the header, and
/// feeds PCM16 to [VoicePlayer] at 24 kHz — Live's own downstream rate, so
/// nothing resamples. Delivery presets are not in the clip: it is the mouth,
/// not the session.
library;

import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

import 'player.dart';
import 'protocol.dart' show voiceAssistantOutputSampleRateV1;

/// Where the minted clip for [voiceName] is bundled.
String voicePreviewAssetV1(String voiceName) => 'assets/voices/$voiceName.wav';

class PcmClipV1 {
  final Uint8List pcm;
  final int sampleRate;
  const PcmClipV1({required this.pcm, required this.sampleRate});
}

int _u32(Uint8List bytes, int offset) =>
    bytes[offset] |
    (bytes[offset + 1] << 8) |
    (bytes[offset + 2] << 16) |
    (bytes[offset + 3] << 24);

String _ascii(Uint8List bytes, int offset, int length) =>
    String.fromCharCodes(bytes.sublist(offset, offset + length));

/// PCM16 little-endian mono from a canonical WAV, or null when it is not one.
PcmClipV1? wavToPcmV1(Uint8List wav) {
  if (wav.length < 44) return null;
  if (_ascii(wav, 0, 4) != 'RIFF' || _ascii(wav, 8, 4) != 'WAVE') return null;
  var offset = 12;
  var sampleRate = voiceAssistantOutputSampleRateV1;
  Uint8List? pcm;
  while (offset + 8 <= wav.length) {
    final id = _ascii(wav, offset, 4);
    final size = _u32(wav, offset + 4);
    final start = offset + 8;
    final end = start + size > wav.length ? wav.length : start + size;
    if (id == 'fmt ' && end - start >= 16) {
      final rate = _u32(wav, start + 4);
      if (rate > 0) sampleRate = rate;
    } else if (id == 'data') {
      pcm = Uint8List.sublistView(wav, start, end);
    }
    offset = start + size + (size.isOdd ? 1 : 0);
  }
  if (pcm == null) return null;
  return PcmClipV1(pcm: pcm, sampleRate: sampleRate);
}

/// A canonical PCM16 mono WAV. The mint script writes this shape; tests
/// build the same bytes so the decoder is proved against a known header.
Uint8List pcmToWavV1(
  Uint8List pcm, {
  int sampleRate = voiceAssistantOutputSampleRateV1,
}) {
  final wav = Uint8List(44 + pcm.length);
  final view = ByteData.sublistView(wav);
  void write(int offset, String text) {
    for (var i = 0; i < text.length; i++) {
      wav[offset + i] = text.codeUnitAt(i);
    }
  }

  write(0, 'RIFF');
  view.setUint32(4, 36 + pcm.length, Endian.little);
  write(8, 'WAVE');
  write(12, 'fmt ');
  view.setUint32(16, 16, Endian.little);
  view.setUint16(20, 1, Endian.little);
  view.setUint16(22, 1, Endian.little);
  view.setUint32(24, sampleRate, Endian.little);
  view.setUint32(28, sampleRate * 2, Endian.little);
  view.setUint16(32, 2, Endian.little);
  view.setUint16(34, 16, Endian.little);
  write(36, 'data');
  view.setUint32(40, pcm.length, Endian.little);
  wav.setRange(44, 44 + pcm.length, pcm);
  return wav;
}

/// Plays one minted timbre at a time through the device speaker.
///
/// Tapping the same voice again stops it. A second voice replaces the first.
/// [dispose] releases the speaker so a later call can own it.
class VoicePreviewPlayer extends ChangeNotifier {
  VoicePreviewPlayer({
    VoicePlayer Function()? createPlayer,
    Future<Uint8List> Function(String voiceName)? loadClip,
  }) : _createPlayer = createPlayer ?? PcmVoicePlayer.new,
       _loadClip =
           loadClip ??
           ((voiceName) async {
             final data = await rootBundle.load(voicePreviewAssetV1(voiceName));
             return data.buffer.asUint8List(
               data.offsetInBytes,
               data.lengthInBytes,
             );
           });

  final VoicePlayer Function() _createPlayer;
  final Future<Uint8List> Function(String voiceName) _loadClip;
  VoicePlayer? _player;
  String? _playing;
  int _generation = 0;

  String? get playing => _playing;

  Future<void> hear(String voiceName) async {
    if (_playing == voiceName) {
      await stop();
      return;
    }
    await stop();
    final generation = ++_generation;
    _playing = voiceName;
    notifyListeners();
    Uint8List bytes;
    try {
      bytes = await _loadClip(voiceName);
    } on Object {
      if (generation != _generation) return;
      _playing = null;
      notifyListeners();
      return;
    }
    if (generation != _generation) return;
    final clip = wavToPcmV1(bytes);
    if (clip == null || clip.pcm.isEmpty) {
      _playing = null;
      notifyListeners();
      return;
    }
    final player = _player ??= _createPlayer();
    await player.configure(clip.sampleRate);
    if (generation != _generation) return;
    player.write(clip.pcm);
    unawaited(
      player.drain().then((_) {
        if (generation != _generation) return;
        _playing = null;
        notifyListeners();
      }),
    );
  }

  Future<void> stop() async {
    _generation++;
    _playing = null;
    notifyListeners();
    final player = _player;
    if (player == null) return;
    await player.interrupt();
  }

  @override
  void dispose() {
    _generation++;
    _playing = null;
    final player = _player;
    _player = null;
    if (player != null) unawaited(player.close());
    super.dispose();
  }
}
